src/expr/type_node.cpp

```diff
@@ -83,46 +83,81 @@ bool TypeNode::isInterpretedFinite()
+      // as finite. If it is not set, we treat them implicitly as infinite
+      // sorts (that is, their cardinality is not constrained to be finite).
+      isInterpretedFinite = options::finiteModelFind();
+    }
+    else if (isBitVector() || isFloatingPoint())
     {
       isInterpretedFinite = true;
     }
-    else if (options::finiteModelFind())
+    else if (isDatatype())
     {
-      if( isSort() ){
-        isInterpretedFinite = true;
-      }else if( isDatatype() ){
-        TypeNode tn = *this;
-        const Datatype& dt = getDatatype();
-        isInterpretedFinite = dt.isInterpretedFinite(tn.toType());
-      }else if( isArray() ){
-        isInterpretedFinite =
-            getArrayIndexType().isInterpretedFinite()
-            && getArrayConstituentType().isInterpretedFinite();
-      }else if( isSet() ) {
-        isInterpretedFinite = getSetElementType().isInterpretedFinite();
+      TypeNode tn = *this;
+      const Datatype& dt = getDatatype();
+      isInterpretedFinite = dt.isInterpretedFinite(tn.toType());
+    }
+    else if (isArray())
+    {
+      TypeNode tnc = getArrayConstituentType();
+      if (!tnc.isInterpretedFinite())
+      {
+        // arrays with consistuent type that is infinite are infinite
+        isInterpretedFinite = false;
       }
-      else if (isFunction())
+      else if (getArrayIndexType().isInterpretedFinite())
       {
+        // arrays with both finite consistuent and index types are finite
         isInterpretedFinite = true;
-        if (!getRangeType().isInterpretedFinite())
-        {
-          isInterpretedFinite = false;
-        }
-        else
+      }
+      else
+      {
+        // If the consistuent type of the array has cardinality one, then the
+        // array type has cardinality one, independent of the index type.
+        isInterpretedFinite = tnc.getCardinality().isOne();
+      }
+    }
+    else if (isSet())
+    {
+      isInterpretedFinite = getSetElementType().isInterpretedFinite();
+    }
+    else if (isFunction())
+    {
+      isInterpretedFinite = true;
+      TypeNode tnr = getRangeType();
+      if (!tnr.isInterpretedFinite())
+      {
+        isInterpretedFinite = false;
+      }
+      else
+      {
+        std::vector<TypeNode> argTypes = getArgTypes();
+        for (unsigned i = 0, nargs = argTypes.size(); i < nargs; i++)
         {
-          std::vector<TypeNode> argTypes = getArgTypes();
-          for (unsigned i = 0, nargs = argTypes.size(); i < nargs; i++)
+          if (!argTypes[i].isInterpretedFinite())
           {
-            if (!argTypes[i].isInterpretedFinite())
-            {
-              isInterpretedFinite = false;
-              break;
-            }
+            isInterpretedFinite = false;
+            break;
           }
         }
+        if (!isInterpretedFinite)
+        {
+          // similar to arrays, functions are finite if their range type
+          // has cardinality one, regardless of the arguments.
```

Comment:
then why not move this block to before checking the arguments and only check them if range is not of size 1?

Reply:
Because it is more expensive to compute the cardinality of type. This calls getCardinality() only if necessary.

Reply:
A further note here: we could implement a dedicated function TypeNode::isCardinalityOne, but I didnt want the changes in this PR to become too large :)

Reply:
It'd be nice to open an issue then